### DISCUSSION TITLE
Get X11 Window ID of a winit window (to make it click-through)

### DISCUSSION THREAD

**Discussion by hir0xygen** [rust-windowing/winit#3953] [Q&A] [answered]
Hey,
I'm currently trying to port a C++ project to Rust, and I need to let mouse and keyboard events be let through my application's window in specific cases.
I understand that this is not possible with winit itself, but from my understanding I could do this by directly interacting with the X11 API to make the window created by winit click-through, though I'm struggling to get the X11 window ID of my winit window.
What's the optimal way to get that?
Thanks in advance for any replies,
hir0xy

EDIT: I apologize, I've completely forgotten to mention, I'm using winit 0.29.15 as I'm unable to compile with 0.30.5 (getting platform unsupported error for some reason), it's the newest version I was able to compile with.

**Comment by kchibisov** [comment]
`Window::set_cursor_hittest` is what you need?

**Reply by hir0xygen** [reply]
I'm not sure whether it's what I was looking for, as my window still won't pass clicks through to the window below it with that set to false.

EDIT: Please also check the edit I appended to my original post, is it just not working on 0.29.15?

This is my window creation function as reference:
```rs
fn create_window(title: &str) -> Result<(Window, EventLoop<()>), Box<dyn Error>> {
  log::debug!("Creating window and event loop");
  let event_loop = EventLoop::new()?;
  event_loop.set_control_flow(ControlFlow::Poll);

  let size = event_loop.primary_monitor().map(|monitor| {
    let monitor_size = monitor.size();
    (monitor_size.width, monitor_size.height)
  });


  let window = WindowBuilder::new()
    .with_title(title)
    .with_transparent(true)
    .with_decorations(false)
    .with_inner_size(PhysicalSize::new(size.map(|(w, _)| w).unwrap_or(WIDTH), size.map(|(_, h)| h).unwrap_or(HEIGHT)))
    .with_resizable(false)
    .with_window_level(WindowLevel::AlwaysOnTop)
    .build(&event_loop)?;
  
  window.set_cursor_hittest(false).unwrap();

  Ok((window, event_loop))
}
```

**Reply by kchibisov** [reply]
If you have a chance to test on e.g. Wayland that would be nice, since on Wayland it should work.

In general, what you do is kind of correct, winit could be just bugged on X11 with that thing.

The `XID` of the window is inside the `raw-window-handle`'s window property, so you can use it for now or check wither latest winit works/fix what we have internally for `hittest`.

**Reply by hir0xygen** [reply]
Was unable to test on Cinnamon Wayland (I usually run Cinnamon X11) due to some unrelated issues, but installed KDE to test there, as I know it's what a lot of my programs users might end up running themselves.

Click-through is indeed working properly on Wayland, and since only the X11 side is broken, I should be able to work around it by setting the window click-through via X11 directly.

Though while on Wayland, I've also found some issues that I can't think of an easy way around currently:
- Transparency does not work
- My way of setting the window to the primary monitor's size, as shown in the code I've posted, does not work, since it's unsupported on Wayland, do you have any ideas what I could do instead? I'd be fine with just prompting the user to select one of the available displays, maybe

Another issue being that `.with_window_level(WindowLevel::AlwaysOnTop)` does not work, but since this is not supported on Wayland it's just something I'll have to deal with some other way. (KDE for example has a feature that allows this to still be realized)

**Reply by kchibisov** [reply, marked as answer]
>Click-through is indeed working properly on Wayland, and since only the X11 side is broken, I should be able to work around it by setting the window click-through via X11 directly.

Then I'd suggest to fix the `X11` impl inside the winit.

>Transparency does not work

You should draw something transparent, though, softbuffer won't do it.

>My way of setting the window to the primary monitor's size, as shown in the code I've posted, does not work, since it's unsupported on Wayland, do you have any ideas what I could do instead? I'd be fine with just prompting the user to select one of the available displays, maybe

You need `ext_layer_shell` on Wayland for that which winit is not supporting. That covers both _on-top_ and span monitor thing.
